drivers/w3s.go

```diff
@@ -0,0 +1,340 @@
+	mu      sync.Mutex
+}
+
+func newRootCar() *rootCar {
+	return &rootCar{
+		root: newDir(),
+		dag:  merkledag.NewDAGService(bserv.New(blockstore.NewBlockstore(dssync.MutexWrap(ds.NewMapDatastore())), nil)),
+	}
+}
+
+type W3sOS struct {
+	ucanKey   string
+	ucanProof string
+	dirPath   string
+	pubId     string
+}
+
+type W3sSession struct {
+	os *W3sOS
+}
+
+func NewW3sDriver(ucanKey, ucanProof, dirPath, pubId string) *W3sOS {
+	return &W3sOS{
+		ucanKey:   ucanKey,
+		ucanProof: ucanProof,
+		dirPath:   dirPath,
+		pubId:     pubId,
+	}
+}
+
+func (ostore *W3sOS) NewSession(filename string) OSSession {
+	if filename != "" {
+		return nil
+	}
+	session := &W3sSession{
+		os: ostore,
+	}
+	return session
+}
+
+func (ostore *W3sOS) UriSchemes() []string {
+	return []string{}
+}
+
+func (ostore *W3sOS) Description() string {
+	return "Web3 Storage driver."
+}
+
+func (session *W3sSession) OS() OSDriver {
+	return session.os
+}
+
+func (session *W3sSession) EndSession() {
+	// no op
+}
+
+func (session *W3sSession) ListFiles(ctx context.Context, cid, delim string) (PageInfo, error) {
+	return nil, ErrNotSupported
+}
+
+func (session *W3sSession) ReadData(ctx context.Context, name string) (*FileInfoReader, error) {
+	return nil, ErrNotSupported
+}
+
+func (session *W3sSession) IsExternal() bool {
+	return false
+}
+
+func (session *W3sSession) IsOwn(url string) bool {
+	// not used anywhere
+	return false
+}
+
+func (session *W3sSession) GetInfo() *OSInfo {
+	return nil
+}
+
+func (session *W3sSession) getAbsolutePath(name string) string {
+	// not supported
+	return ""
+}
+
+func (session *W3sSession) SaveData(ctx context.Context, name string, data io.Reader, meta map[string]string, timeout time.Duration) (string, error) {
+	if timeout <= 0 {
+		timeout = w3SDefaultSaveTimeout
+	}
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	filePath, err := toFile(data)
+	if err != nil {
+		return "", err
+	}
+	defer deleteFile(filePath)
+
+	carPath, fileCid, err := ipfsCarPack(ctx, filePath)
+	if err != nil {
+		return "", err
+	}
+	defer deleteFile(carPath)
+
+	carCid, err := w3StoreCar(ctx, carPath)
+	if err != nil {
+		return "", err
+	}
+
+	rCar := session.os.getRootCar()
+	if err = rCar.addFile(ctx, session.os.dirPath, name, fileCid, carCid); err != nil {
+		return "", err
+	}
+
+	return fileCid, nil
+}
+
+func (rc *rootCar) addFile(ctx context.Context, dirPath, filename, fileCid, carCid string) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	rc.carCids = append(rc.carCids, carCid)
+
+	// split path by "/", ignore empty strings
+	dirPaths := strings.FieldsFunc(dirPath, func(c rune) bool { return c == '/' })
+
+	newRoot, err := rc.addFileToDag(ctx, rc.root, dirPaths, filename, fileCid)
+	if err != nil {
+		return err
+	}
+	rc.root = newRoot
+
+	return nil
+}
+
+func (rc *rootCar) addFileToDag(ctx context.Context, n *merkledag.ProtoNode, dirPaths []string, filename, fileCid string) (*merkledag.ProtoNode, error) {
```

Comment:
Nice. I think the recursive logic here makes sense. If I'm following correctly, it works like this:

- We start at the root path and have a level of recursion for each path in `dirPaths`
- At each level, if the path does not represent a child node:
    - If there is already a node for the path, we return it
    - If there is not a node for the path, we create an empty dir node
- If the path does represent a child node, create it with the CID
- As we recurse back up through each level we have to update the link (i.e. the CID) of the node at this level if there was an update to one of its children

However, I wonder what this logic was based off. Do we know that this is the typical logic used to build a directory and was there a reference used? I was taking a look at [this go-car create command](https://github.com/ipld/go-car/blob/268b65bab2b7f45aa05494b943992f5b9bbab75e/cmd/car/create.go#L106) and noticed that there is a [builder.BuildUnixFSRecursive()](https://github.com/ipfs/go-unixfsnode/blob/d3fccee9b573daccbb1b0b6f9972f598d04c321e/data/builder/directory.go#L24) function in https://github.com/ipfs/go-unixfsnode/. At a glance, it looks like it might do something similar to the logic you implemented here although it seems to assume that the files are available in the local filesystem. Do you think using that function could simplify our implementation here or would it not fit our needs?

Reply:
I took at look at the go-car lib and I think we can't use `builder.BuildUnixFSRecursive()` (or any other function from the `ipfs/go-unixfsnode` library).

The function `builder.BuildUnixFSRecursive()` does something different, it creates DAG from the existing directory structure in the local FS. We do not have any directory structure in our local FS and I think we don't want to create it.

Saying that, I add a comment explaining the logic behind the `addFileDag()` function. It looks difficult, but when you realize that all it does is to use the DFS algorithm to create nodes and leaves it is a little more digestible.

Reply:
Ah bummer okay!